### DISCUSSION TITLE
Having troubles setting up sliding environment with constant interaction force

### DISCUSSION THREAD

**Discussion by lucky-luke-98** [google-deepmind/mujoco#1925] [Asking for Help] [answered]
Hello everybody,

I am new to the mujoco simulations and I am having some troubles setting up an experiment:
The goal is, that a robotic arm with a wiper end-effector slides over a 3D bridge with constant force (wiping task).

Some specifications about the objects:
<b>wiper:</b>
- geometry: cylindric
- material: plastic, coated with a 0.5cm layer of felt
- stiffness: medium

<b>bridge:</b>
- geometry: convex (similar to convex bridge)
- material: similar to metal
- stiffness: high

I added a site to the end of the wiper that has a force sensor attached to it, to measure the contact force. In order to make the wiper slide over the bridge, I also defined a motion planning law that acts on a given trajectory and the necessary orientations for the bridge's geometry. The frequency of this motion planning is 30 Hz while I have also specified an impedance controller that acts on the motion plannings action in 600 Hz (same frequency as for the mujoco simulation).

In the XML-files of my simulation I specified the objects contact behaviour as follows:

wiper:
```xml
<body name="wiper" pos="0 0 0.107" quat="1 0 0 0">
    <site name="ft_frame" pos="0 0 0" size="0.0001" rgba="1 1 0 0" type="sphere" group="1"/>
    <inertial pos="0.00 0.001 -0.018" mass="0.709" diaginertia="55e-8 54e-8 53e-8"/>
    <geom type="mesh" pos="0 0 0.0" name="wiping_surface_vis" mesh="wipe_eef" contype="1" group="1" solimp="0.05 0.9 0.1 0.1 6" solref="0.05 0.5" priority="1" condim="4" friction="0.0001 0.0005 0.0001" rgba="0.25 0.25 0.25 1" />
          
    <!-- < For visusalization > -->
    <body name="eef" pos="0 0 0.0536" quat="1 0 0 0">
        <inertial pos="0 0 0" mass="0.0000001" diaginertia="0 0 0" />
        <geom type="cylinder" pos="0 0 0" size="0.025 0.00001" rgba="0 1 0 0" group="1"/>
        <site name="grip_site" pos="0 0 0" size="0.001 0.001 0.001" rgba="1 0 0 0.5" type="sphere" group="1"/>
         ...
    </body>
</body>

...
<force name="eef_force" site="grip_site"/>
```

bridge:
```xml
<body name="object">
    <geom name="bridge" pos="0.5 0 .97" mesh="bridge_mesh" type="mesh" density="50" group="0" condim="4" solimp="0.75 0.99 0.001" solref="0.02 0.2" friction="0.001 0.01 0.01"/>
</body>
```

Note, that these snippets are only parts of the XML file specifications that specify the contacts.
The key notes here are, that I have chosen a low friction for the wiper, which has higher priority than the bridge. I am specifying a condim of 4 since I do not want rolling behaviour (just sliding).

___

### Expected behaviour:
I expect the wiper to slide over the bridge with constant measured force. 

### Current outcome:
Although I am experiencing a constant velocity over the bridge (due to my motion planning), I am also experiencing a weird skipping motion of the wiper. This skipping motion leads the wiper to lift from the bridge which results in no contact force between the two objects for a slight moment. Then the wiper is getting in contact again leading to a force. This behaviour can be seen in the next image, where I have plotted the measured force of the sensor. In blue I plotted the overall changes of the force in the frequency of the simulation physics dt while I plotted the general reading times of the force sensor (frequency of my motion planning law) in red.
![image width="0.5"](https://github.com/user-attachments/assets/1b418ef8-bf53-459f-a65a-d498ac5ae664)
Hence, the measured force is either the expected force or 0.0.

### My question:
Is that problem originating from my simulation specifications? I mean, should I expect this kind of contact behaviour since my specifications of the simulation are bad chosen? 
Another possible reason for this might be the control theory (impedance controller) but before I check this, I want to be certain about the choices of my simulation.

Thanks in advance!

**Comment by bmaxdk** [comment]
You may want to tune friction. Start by slightly increasing the friction values for both the wiper and bridge to see if that helps stabilize the contact. Your current friction values : `friction="0.0001 0.0005 0.0001"` for the wiper and `friction="0.001 0.01 0.01"` for the bridge are quite low. If the friction is too low, it may not provide enough resistance to the sliding motion, causing the wiper to occasionally lose contact. You might want to experiment with slightly higher friction coefficients.

Also try adjust Adjust `solimp` and `solref`. Those parameters define the softness of the contact and the velocity at which the contact forces are computed. You want to determin how the contact interaction behaves. 

Can you also try `solimp="0.8 0.99 0.001`  and  Adjust `solref` to control the damping effect `solref="0.02 0.1"`?

Hope it works

**Reply by lucky-luke-98** [reply]
Thanks for your reply! I was not aware of the effect a too low friction might have. 
Changing the values as advised by you, did not really change the behaviour that I am experiencing. I guess it rather is a problem of my specified impedance gains for the impedance controller.
But it's a good thing to get my chosen values for friction and stiffness verified by you.

**Comment by lucky-luke-98** [comment, marked as answer]
<b>For your interest:</b> 
I found out the solution to my problem. Above, you can see that I am changing the attributes of the body called <i>wiper</i>. However, since I am incorporating a force sensor, the <i>wiper</i> will never be in contact with the <i>bridge</i>. The body that interacts with the <i>bridge</i> is the geom <i>cylinder</i> which is part of the eef.
Therefore, I have to change those attributes. 

After changing this, I experienced much better behavior of my simulation.
